### ISSUE TITLE
SIGWINCH while blocking read causes the read to stop blocking

### ISSUE THREAD

**Issue by leath-dub** [mame/xterm-pty#41]
I found this issue when running the lua interpreter with xterm-pty and xterm.js. It seems the culprit is this code: https://github.com/mame/xterm-pty/blob/main/emscripten-pty.js#L100.

This will essentially mean any blocking read is discarded on any signal, and whenever you resize the terminal it will discard the read. I have a hack fix that is a special case for SIGWINCH, I understand the better way is to only stop when the handler is called as the comment suggests, but this is pretty important for the time being I would say. The new handler would be simply:

```js
      handlerSignal = PTY.onSignal(
        (sig) => {
          if (PTY_signalNameToCode[sig] != 28 /* SIGWINCH */) {
            resolve(1 /* interrupted */)
          }
        }
      );
```

**Comment by mame** [comment]
Unfortunately, the unblocking behavior of SIGWINCH is actually needed in the vim case. The vim is always blocked waiting for input. When the terminal is resized, it need to redraw the screen. So vim expects the blocking read is interrupted by SIGWINCH.

So, I think I can't just merge your hack as is. I wonder if there is a better way. Maybe we could put `#if IGNORE_SIGWINCH` and ignore SIGWINCH only when `emcc -sIGNORE_SIGWINCH`?

**Comment by leath-dub** [comment]
> Unfortunately, the unblocking behavior of SIGWINCH is actually needed in the vim case. The vim is always blocked waiting for input. When the terminal is resized, it need to redraw the screen. So vim expects the blocking read to be interrupted by SIGWINCH.
> 
> So, I think I can't just merge your hack as is. I wonder if there is a better way. Maybe we could put `#if IGNORE_SIGWINCH` and ignore SIGWINCH only when `emcc -sIGNORE_SIGWINCH`?

Right now I just have the patched function as post.js, I think that is fine, I could right some note about it though. I think ultimately the issue is that it needs to block if the default signal handler was called: I assume vim does not use the default signal handler.

**Comment by mame** [comment]
> I think ultimately the issue is that it needs to block if the default signal handler was called

@leath-dub By #44, IO-blocking operation is now interrupted only when non-default signal handler is called. So I believe this issue is fixed. If you have time, try the following version of emscripten-pty.js.

https://github.com/mame/xterm-pty/blob/a9b36aade54dde03af9f8be0dab4576fcebe7563/emscripten-pty.js

**Comment by leath-dub** [comment]
> > I think ultimately the issue is that it needs to block if the default signal handler was called
> 
> [@leath-dub](https://github.com/leath-dub) By [#44](https://github.com/mame/xterm-pty/pull/44), IO-blocking operation is now interrupted only when non-default signal handler is called. So I believe this issue is fixed. If you have time, try the following version of emscripten-pty.js.
> 
> https://github.com/mame/xterm-pty/blob/a9b36aade54dde03af9f8be0dab4576fcebe7563/emscripten-pty.js

Sorry for the delay here, thanks this works great !!!

**Comment by mame** [comment]
Just FYI: I have published xterm-pty 0.11.0 with this fix. Thanks!
